BUG: Fix loading files from S3 with # characters in URL (GH25945)

swt2c opened the pull request:
This fixes loading files with URLs such as s3://bucket/key#1.csv.  The part
from the # on was being lost because it was considered to be a URL fragment.
The fix disables URL fragment parsing as it doesn't make sense for S3 URLs.

- [x] closes #25945 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

swt2c commented:
I believe the CI failures are due to my added test, which is causing s3fs to be imported due to importing the `_strip_schema` function from `pandas.io.s3`.  This was really the only way I could see to add a test for this fix, other than try to load something from an actual S3 bucket.  Is it okay to add s3fs as a test dependency, and if so, how do I do that?

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -356,6 +356,7 @@ I/O
 - Bug in :func:`read_hdf` not properly closing store after a ``KeyError`` is raised (:issue:`25766`)
 - Bug in ``read_csv`` which would not raise ``ValueError`` if a column index in ``usecols`` was out of bounds (:issue:`25623`)
 - Improved :meth:`pandas.read_stata` and :class:`pandas.io.stata.StataReader` to read incorrectly formatted 118 format files saved by Stata (:issue:`25960`)
+- Fixed bug in loading objects from S3 that contain # characters in the URL (:issue:`25945`)
```
use double-back ticks on the ``#``. you can say this is in ``read_*`` routines such as ``:func:read_csv``

swt2c replied:
Fixed.

jreback commented on pandas/tests/io/test_s3.py:
```diff
@@ -27,3 +28,8 @@ def test_streaming_s3_objects():
         read_csv(body)
+
+
+def test_parse_s3_url_with_pound_sign():
```
this is not a useful test, you need to mock the routines as above

swt2c replied:
Thanks for the review.  I have to disagree with you about this not being a useful test.  It tests the buggy function, fails without the fix and passes with the fix.  Can you provide any more hints about how I would mock this as you suggested?  I can't see how to do it.

jreback replied:
look at any tests in pandas/tests/io//parser/test_network.py, you need to use the s3_resource fixture.

swt2c replied:
Thanks, I believe I figured it out now.  :)  `moto` seems very nice, I hadn't heard of it before.

jreback reviewed (changes requested):
lgtm. small comments, ping on green.

swt2c committed:
BUG: Fix loading files from S3 with # characters in URL (GH25945)

This fixes loading files with URLs such as s3://bucket/key#1.csv.  The part
from the # on was being lost because it was considered to be a URL fragment.
The fix disables URL fragment parsing as it doesn't make sense for S3 URLs.

WillAyd commented:
Thanks @swt2c
